src/chkstat.cpp

```diff
@@ -21,1365 +21,1246 @@
-                {
-                    extline = 1;
-                    break;
-                }
-                if (!inpart)
-                {
-                    inpart = 1;
-                    if (pcnt == 3)
-                        break;
-                    part[pcnt] = p;
-                }
+                std::cerr << ctx.subpath << ": has unexpected owner (" << ctx.status.st_uid << "). refusing to correct due to unknown integrity." << std::endl;
+                return false;
             }
-            if (extline)
+            else
             {
-                if (!last)
-                {
-                    BAD_LINE();
-                    continue;
-                }
-                if (!strncmp(p, "+capabilities ", 14))
-                {
-                    if (have_fscaps != 1)
-                        continue;
-                    p += 14;
-                    caps = cap_from_text(p);
-                    if (caps)
-                    {
-                        cap_free(last->caps);
-                        last->caps = caps;
-                    }
-                    continue;
-                }
-                BAD_LINE();
-                continue;
+                const auto path = getPathFromProc(pathfd);
+                std::cerr << ctx.subpath << ": on an insecure path - " << path << " has different non-root owner who could tamper with the file." << std::endl;
+                return false;
             }
-            if (inpart)
-                pcnt++;
-            if (pcnt != 3)
+        }
+
+        if (ctx.status.isLink())
+        {
+            // If the path configured in the permissions configuration is a symlink, we don't follow it.
+            // This is to emulate legacy behaviour: old insecure versions of chkstat did a simple lstat(path) as 'protection' against malicious symlinks.
+            if (is_final_path_element || ++link_count >= 256)
+                // TODO: would an excess link count not warrant a complaint?
+                return false;
+
+            // Don't follow symlinks owned by regular users.
+            // In theory, we could also trust symlinks where the owner of the target matches the owner
+            // of the link, but we're going the simple route for now.
+            if (ctx.status.st_uid && ctx.status.st_uid != m_euid)
             {
-                BAD_LINE();
-                continue;
+                const auto path = getPathFromProc(pathfd);
+                std::cerr << ctx.subpath << ": on an insecure path - " << path << " has different non-root owner who could tamper with the file." << std::endl;
+                return false;
             }
-            part[3] = part[2];
-            part[2] = strchr(part[1], ':');
-            if (!part[2])
-                part[2] = strchr(part[1], '.');
-            if (!part[2])
+
+            link.resize(PATH_MAX);
+            const auto len = ::readlinkat(pathfd.get(), "", &link[0], link.size() - 1);
```

Comment:
C++ doesn't require a NUL byte in the end, so no need to subtract 1 from the string length (also two lines below).